Created public get_topic_list() function for use in other scripts.

EliteMasterEric opened the pull request:
Resolves issue #946.

The only function in rostopic for listing the available topics is built exclusively for the command line interface; this adds a public function for other applications that want a list of topics.

Usage:
```
import rostopic
publishers, subscribers = rostopic.get_topic_list()
for topic, type, node_count in publishers:
    print("Publisher: %s, [%s] (%d nodes)" % (topic, type, node_count))
for topic, type, node_count in subscribers:
    print("Subscriber: %s, [%s] (%d nodes)" % (topic, type, node_count))
```

Commit:
Implemented rate-control-topic and rate-control-max-delay.

Commit:
Made rate control comparisons compare to time publisher rather than wall time.

Commit:
Added a check and a warning if the header is not the first in the message,

and removed an unnecessary import.

EliteMasterEric committed:
Some quick formatting changes to simplify the final pull reuqest.

EliteMasterEric committed:
Why is there little whitespace changes.

EliteMasterEric committed:
Tabs to spaces, and other whitespace fixes.

EliteMasterEric committed:
This is the last whitespace fix I swear.

Commit:
Removed unused function parameters.

Commit:
Fixed a bug preventing code from building after commit ecfb85f.

dirk-thomas committed:
Merge branch 'kinetic-devel' into kinetic-devel

khallenbeck-dsi committed:
Aligned print messages for rosbag PAUSED/DELAYED/RUNNING

dirk-thomas commented:
After extracting the print-agnostic part of the existing functionality as you did in the current patch please also update the existing code to use this new function (rather than duplicating this logic). Please also consider adding a test for the new function.

EliteMasterEric commented:
@dirk-thomas Changes have been made, removing redundancy and adding a test to the rostopic test script.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -1155,6 +1155,31 @@ def topic_type(t, topic_types):
+    topic_types = _master_get_topic_types(master)
+    
+    pubs, subs, _ = state
+    publishers = []
```
This line is overwriting the function parameter.

Same below for `subscribers`.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -1155,6 +1155,31 @@ def topic_type(t, topic_types):
+    publishers = []
+    if publishers:
+        for topic, nodes in pubs:
+            publishers.append(topic, topic_type(topic, topic_types), len(nodes))
```
`list.append()` only accepts a single argument.

Same below.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -1202,14 +1227,7 @@ def _rostopic_list(topic, verbose=False,
-            topic_ns = rosgraph.names.make_global_ns(topic)        
-            subs = (x for x in subs if x[0] == topic or x[0].startswith(topic_ns))
-            pubs = (x for x in pubs if x[0] == topic or x[0].startswith(topic_ns))
+        pubs, subs = get_topic_list()
```
The new function returns different information than `getSystemState` before. I doubt that all following code continues to work as is.

dirk-thomas reviewed (commented):
Please make sure to run all code parts you have changed in all possible variations to make sure the patch actually works (which is currently not the case).

EliteMasterEric committed:
Created public get_topic_list() function for use in other scripts.

The existing rostopic_list functions output straight to print() calls, so they are only functional for the CLI. Resolves issue #946.
Added new public function to tests.

Updated functions to use new public function to remove redundancy.
